Branch-TestBed/Branch-SDK-Tests/BNCServerInterfaceTests.m

```diff
@@ -19,176 +23,327 @@ @interface BNCServerInterfaceTests : XCTestCase
-        return [request.URL.query rangeOfString:@"app_id=non_branch_key"].location != NSNotFound;
-    }] queue:[OCMArg any] completionHandler:[OCMArg any]];
-    
-    // Make the request
+    XCTestExpectation* expectation = [self expectationWithDescription:@"NSURLSessionDataTask completed"];
+
+    [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest *request) {
+        // We're not sending a request, just verifying a "branch_key=key_foo" is present.
+        XCTAssertTrue([request.URL.query rangeOfString:@"app_id=non_branch_key"].location != NSNotFound, @"Branch App ID not added");
+        [expectation fulfill];
+        return [request.URL.query rangeOfString:@"app_id=key_foo"].location != NSNotFound;
+    } withStubResponse:^OHHTTPStubsResponse *(NSURLRequest *request) {
+        NSDictionary* dummyJSONResponse = @{@"key": @"value"};
+        return [OHHTTPStubsResponse responseWithJSONObject:dummyJSONResponse statusCode:200 headers:nil];
+    }];
+
     [serverInterface getRequest:nil url:@"http://foo" key:@"non_branch_key" callback:NULL];
-    
-    [urlConnectionMock verify];
+
+    [self waitForExpectationsWithTimeout:5.0 /* 5 seconds */ handler:nil];
 }
 
 
+
+
 #pragma mark - Retry tests
+//==================================================================================
+//TEST 03
+//This test simulates a poor network, with three failed GET attempts and one final success,
+// for 4 connections.
 
 - (void)testGetRequestAsyncRetriesWhenAppropriate {
-    BNCServerInterface *serverInterface = [[BNCServerInterface alloc] init];
-    serverInterface.preferenceHelper = [[BNCPreferenceHelper alloc] init];
-    id urlConnectionMock = OCMClassMock([NSURLConnection class]);
-    
-    // Specify retry count as 3
-    serverInterface.preferenceHelper.retryCount = 3;
+  
+  //Set up nsurlsession and data task, catching response
+  BNCServerInterface *serverInterface = [[BNCServerInterface alloc] init];
+  serverInterface.preferenceHelper = [[BNCPreferenceHelper alloc] init];
+  serverInterface.preferenceHelper.retryCount = 3;
 
-    // 3 retries means 4 total requests
-    [self expectSendAsyncRequestForBranchError:urlConnectionMock times:4];
-    
-    // Make the request
-    XCTestExpectation *getRequestExpectation = [self expectationWithDescription:@"GET Request Expectation"];
-    [serverInterface getRequest:nil url:@"http://foo" key:@"key_foo" callback:^(BNCServerResponse *response, NSError *error) {
-        XCTAssertNotNil(error);
+  XCTestExpectation* successExpectation = [self expectationWithDescription:@"success"];
+  
+  __block NSUInteger connectionAttempts = 0;
+  __block NSUInteger failedConnections = 0;
+  __block NSUInteger successfulConnections = 0;
+  
+  [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest *request) {
+    BOOL foundBranchKey = [request.URL.query rangeOfString:@"branch_key=key_foo"].location != NSNotFound;
+    XCTAssertEqual(foundBranchKey, TRUE);
```

Comment:
we can probably stub this since this isn't critical for retries - or is there another reason this was needed?